src/pow.cpp

```diff
@@ -197,6 +197,51 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
     return bnNew.GetCompact();
 }
 
+// Called after activation of the protocol-cleanup rule changes, at
+// which time the difficulty adjustment is largely unchecked. For DoS
+// prevention purposes we require that the difficulty adjust by no
+// more than +/- 2x as compared with the difficulties of the last 12
+// blocks. This is enough of a connstraint that any DoS attack is
+// forced to have non-trivial mining costs (e.g. equal to extending
+// the tip by 6 blocks to reduce difficulty by more than a half, work
+// equal to extending the tip by 9 blocks to reduce by more than a
+// quarter, 10.5 times present difficulty to reduce by more than an
+// eigth, etc. To reduce to arbitrary levels requires 12 blocks worth
+// of work at the difficulty of the last valid block.
+bool CheckNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader& block)
+{
+    // Special case for the genesis block
+    if (!pindexLast)
+        return (block.nBits == Params().GenesisBlock().nBits);
+
+    // If look reversed, that is to be expected. We set min to the
+    // largest possible value, and max to the smallest. That way these
+    // will be replaced with actual block values as we loop through
+    // the past 12 blocks.
+    uint256 min = Params().ProofOfWorkLimit().GetCompact();
+    uint256 max = uint256(1);
+
+    // After this loop, min will be half the largest work target of
+    // the past 12 blocks, and max will be twice the smallest.
+    for (int i = 0; i < 12 && pindexLast; ++i, pindexLast = pindexLast->pprev) {
+        uint256 target;
+        target.SetCompact(pindexLast->nBits);
+        uint256 local_min = target >> 1;
+        uint256 local_max = target << 1;
+        if (min > local_min)
+            min = local_min;
+        if (max < local_max)
+            max = local_max;
```

Comment:
Don't know if you care (let me know if, so I can stop pointing these out), but the coding style guidelines over on Bitcoin Core land currently say that you either put brackets or put the code on the same line as the if case, i.e.
```C++
if (min > local_min) min = local_min;
if (max < local_max) max = local_max;
```
or
```C++
if (min > local_min) {
    min = local_min;
}
if (max < local_max) {
    max = local_max;
}
```

Reply:
I try to never do it the first way. But brackets are good practice and I'll make that change.

Reply:
Done.

Reply:
OK. I will point out unbracketed new code when I see it.